File: SPOS/New Codes/round_robin.py
from collections import deque

def round_robin():
    n = int(input("Enter number of processes: "))
    arrival = []
    burst = []
    remaining = []
    completion = [0] * n
    waiting = [0] * n
    turnaround = [0] * n
    response = [-1] * n


    for i in range(n):
            at = int(input(f"Enter arrival time of P{i}: "))
            bt = int(input(f"Enter burst time of P{i}: "))
            arrival.append(at)
            burst.append(bt)
            remaining.append(bt)

    timeQuantum = int(input("Enter time quantum: "))

    currentTime = 0
    q = deque()
    inQueue = [False] * n
    completedProcesses = 0

    
    for i in range(n):
        if arrival[i] <= currentTime:
            q.append(i)
            inQueue[i] = True

    while completedProcesses < n:
        if not q:
            currentTime += 1
            for i in range(n):
                if not inQueue[i] and arrival[i] <= currentTime and remaining[i] > 0:
                    q.append(i)
                    inQueue[i] = True
            continue

        idx = q.popleft() 

        if response[idx] == -1:
            response[idx] = currentTime - arrival[idx]

        if remaining[idx] <= timeQuantum:
            currentTime += remaining[idx]
            remaining[idx] = 0
            completion[idx] = currentTime
            turnaround[idx] = completion[idx] - arrival[idx]
            waiting[idx] = turnaround[idx] - burst[idx]
            completedProcesses += 1
        else:
            currentTime += timeQuantum
            remaining[idx] -= timeQuantum

        
        for i in range(n):
            if not inQueue[i] and arrival[i] <= currentTime and remaining[i] > 0:
                q.append(i)
                inQueue[i] = True

        
        if remaining[idx] > 0:
            q.append(idx)

    print("\n| Process | Completion Time | Waiting Time | Turnaround Time | Response Time |")
    totalWaiting = totalTurnaround = totalCompletion = 0
    for i in range(n):
        print(f"| P{i:<6} | {completion[i]:<15} | {waiting[i]:<12} | {turnaround[i]:<15} | {response[i]:<13} |")
        totalWaiting += waiting[i]
        totalTurnaround += turnaround[i]
        totalCompletion += completion[i]

    print(f"\nAverage Completion Time: {totalCompletion / n:.2f}")
    print(f"Average Waiting Time: {totalWaiting / n:.2f}")
    print(f"Average Turnaround Time: {totalTurnaround / n:.2f}")

if __name__ == "__main__":
    round_robin()

File: SPOS Java Codes/RoundRobin.java
import java.util.*;

public class RoundRobin {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        System.out.print("Enter number of processes: ");
        int n = sc.nextInt();
        
        int[] pid = new int[n];      // Process IDs
        int[] arrival = new int[n];  // Arrival times
        int[] burst = new int[n];    // Burst times
        int[] remaining = new int[n]; // Remaining burst times
        int[] completion = new int[n]; // Completion times
        int[] turnaround = new int[n]; // Turnaround times
        int[] waiting = new int[n];   // Waiting times
        
        // Input process details
        for(int i = 0; i < n; i++) {
            System.out.print("Enter arrival time for process " + (i+1) + ": ");
            arrival[i] = sc.nextInt();
            System.out.print("Enter burst time for process " + (i+1) + ": ");
            burst[i] = sc.nextInt();
            remaining[i] = burst[i];
            pid[i] = i + 1;
        }
        
        System.out.print("Enter time quantum: ");
        int quantum = sc.nextInt();
        
        int currentTime = 0;
        int completed = 0;
        
        // Round Robin scheduling
        while(completed < n) {
            boolean executed = false;
            
            for(int i = 0; i < n; i++) {
                // Process is ready and has remaining burst time
                if(arrival[i] <= currentTime && remaining[i] > 0) {
                    executed = true;
                    
                    if(remaining[i] > quantum) {
                        // Execute for quantum time
                        currentTime += quantum;
                        remaining[i] -= quantum;
                    } else {
                        // Process completes execution
                        currentTime += remaining[i];
                        remaining[i] = 0;
                        completed++;
                        completion[i] = currentTime;
                    }
                }
            }
            
            // If no process was executed in this cycle, increment time
            if(!executed) {
                currentTime++;
            }
        }
        
        // Calculate turnaround and waiting times
        float avgWait = 0, avgTurnaround = 0;
        for(int i = 0; i < n; i++) {
            turnaround[i] = completion[i] - arrival[i];
            waiting[i] = turnaround[i] - burst[i];
            avgWait += waiting[i];
            avgTurnaround += turnaround[i];
        }
        
        // Display results
        System.out.println("\nProcess Table:");
        System.out.println("PID\tArrival\tBurst\tCompletion\tTurnaround\tWaiting");
        for(int i = 0; i < n; i++) {
            System.out.printf("%d\t%d\t%d\t%d\t\t%d\t\t%d\n", 
                pid[i], arrival[i], burst[i], completion[i], turnaround[i], waiting[i]);
        }
        
        System.out.printf("\nAverage Turnaround Time: %.2f\n", (avgTurnaround / n));
        System.out.printf("Average Waiting Time: %.2f\n", (avgWait / n));
        
        sc.close();
    }
}